### PR TITLE
Implement basic support for managed exception handling

### PR DESCRIPTION
**Pull request by sergiy-k**
Implementation of the managed exception handling in this change is by far not yet complete but it is a good starting point that we can build on top of it. Basically this code allows managed exceptions to be thrown and caught. The finally blocks and nested try/catch work too. But re-throwing an exception from a catch block and many other corner cases do not work yet.

In addition, RtlRestoreContext needs to be properly implemented. This change introduces a very simply implementation that works only in those cases where XMM registers are not used in the code that handles the exception so they do not need to be restored. I have created a separate issue to track it (https://github.com/dotnet/coreclr/issues/360).

This change also fixes an issue in JIT where JIT was incorrectly passing arguments in the RCX and RDX registers to the finally and catch funclets.

### CODE REVIEW

**Comment by sergiy-k** [comment]
Here is an example of test code that works:

using System;

class Program
{
    static void A()
    {
        try
        {
            Console.WriteLine("Throwing an exception");
            throw new Exception("Did you catch it?");
        }
        finally
        {
            Console.WriteLine("A.finally()");
        }
    }

    static void B()
    {
        try
        {
            C();
            Console.WriteLine("Failure! Exception has not been thrown.");
        }
        catch (Exception ex)
        {
            Console.WriteLine("Test B: success! Caught exception!\n" + ex.ToString());
        }
        finally
        {
            Console.WriteLine("B.finally()");
        }
    }

    static void C()
    {
        A();
    }

    static void D()
    {
        try
        {
            try
            {
                Console.WriteLine("Throwing an exception");
                throw new Exception("Did you catch it in the right handler?");
            }
            catch (IndexOutOfRangeException e)
            {
                Console.WriteLine("Test D: Failed. Called wrong handler.\n" + e.ToString());
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("Test D: success! Caught exception!\n" + ex.ToString());
        }
    }

    static void Main()
    {
        Console.WriteLine("Testing A");
        try
        {
            A();
            Console.WriteLine("Failure! Exception has not been thrown.");
        }
        catch (Exception ex)
        {
            Console.WriteLine("Test A: success! Caught exception!\n" + ex.ToString());
        }
        finally
        {
            Console.WriteLine("Main.finally().");
        }

        Console.WriteLine("\nTesting B");

        B();

        Console.WriteLine("\nTesting D");

        D();

        Console.WriteLine("\nTesting class TestE");

        TestE.Run();

        Console.WriteLine("Exiting test");
    }

    class TestE
    {
        static void A()
        {
            try
            {
                Console.WriteLine("In A");
                B();
            }
            catch (ArgumentException ae)
            {
                Console.WriteLine("Error! Caught ArgumentException.\n" + ae.ToString());
            }
        }

        static void B()
        {
            try
            {
                Console.WriteLine("In B");
                C();
            }
            finally
            {
                Console.WriteLine("Leaving B");
            }
        }

        static void C()
        {
            Console.WriteLine("In C");
            D();
            Console.WriteLine("Error! A() should not be reached in C()");
            A();
        }

        static void D()
        {
            Console.WriteLine("In D, throwing...");
            throw new Exception("Exception test");
        }

        public static void Run()
        {
            try
            {
                A();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Test C: success! Caught exception!\n" + ex.ToString());
            }
        }
    }
}

**Comment by kangaroo** [comment]
@sergiy-k Awesome!  I've been playing around with this as well, but you beat me to it.  I'll try to TAL on OS X before heading to the airport.

File: src/pal/src/debug/debug.cpp [hunk 1]
```diff
@@ -530,7 +530,27 @@ RtlRestoreContext(
   IN PEXCEPTION_RECORD ExceptionRecord
 )
 {
-    ASSERT("UNIXTODO: Implement this");
+    native_context_t ucontext;
+
```
Comment (kangaroo) [inline src/pal/src/debug/debug.cpp]: This won't work on OSX.  We don't have native_context_t, we probably want CONTEXT_SetThreadContextOnPort here.

**Comment by kangaroo** [comment]
It seems to be mostly working, modulo my comments, except I'm currently ending up in:

```
(lldb) bt
* thread #1: tid = 0x1097ee, 0x000000011b54a7b5, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=EXC_I386_BPT, subcode=0x0)
  * frame #0: 0x000000011b54a7b5
    frame #1: 0x000000011b54774f
    frame #2: 0x0000000101679019 libcoreclr.dylib`CallDescrWorkerInternal + 125 at calldescrworkeramd64.S:902
    frame #3: 0x00000001014ad772 libcoreclr.dylib`CallDescrWorkerWithHandler(pCallDescrData=0x00007fff5fbfe570, fCriticalCall=0) + 482 at callhelpers.cpp:88
    frame #4: 0x00000001014aea0a libcoreclr.dylib`MethodDescCallSite::CallTargetWorker(this=0x00007fff5fbfe790, pArguments=0x00007fff5fbfe718) + 2426 at callhelpers.cpp:627
    frame #5: 0x000000010143da2d libcoreclr.dylib`MethodDescCallSite::Call(this=0x00007fff5fbfe790, pArguments=0x00007fff5fbfe718) + 29 at callhelpers.h:421
    frame #6: 0x0000000101112094 libcoreclr.dylib`RunMain(this=0x00007fff5fbfe8a8, pParam=0x00007fff5fbfe940)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const::'lambda'(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*)::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const + 580 at assembly.cpp:2540
    frame #7: 0x000000010110f3ec libcoreclr.dylib`RunMain(this=0x00007fff5fbfe930, __EXparam=0x00007fff5fbfe940)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*)::Param*) const + 76 at assembly.cpp:2558
    frame #8: 0x000000010110f283 libcoreclr.dylib`RunMain(pFD=0x00000001000e4950, numSkipArgs=1, piRetVal=0x00007fff5fbfea58, stringArgs=0x00007fff5fbfedb0) + 323 at assembly.cpp:2558
    frame #9: 0x000000010110f5b8 libcoreclr.dylib`Assembly::ExecuteMainMethod(this=0x000000010021be88, stringArgs=0x00007fff5fbfedb0) + 328 at assembly.cpp:2666
    frame #10: 0x000000010119238f libcoreclr.dylib`CorHost2::ExecuteAssembly(this=0x00000001002019b8, dwAppDomainId=1, pwzAssemblyPath=0x0000000100215db8, argc=0, argv=0x0000000000000000, pReturnValue=0x00007fff5fbff2fc) + 1167 at corhost.cpp:1309
    frame #11: 0x0000000101004c77 libcoreclr.dylib`ExecuteAssembly(exePath=0x00007fff5fbffcb0, coreClrPath=0x0000000100104c30, appDomainFriendlyName=0x0000000100007d80, propertyCount=4, propertyKeys=0x00007fff5fbff770, propertyValues=0x00007fff5fbff750, argc=0, argv=0x0000000000000000, managedAssemblyPath=0x0000000100104b80, exitCode=0x00007fff5fbff2fc) + 1655 at unixinterface.cpp:185
    frame #12: 0x000000010000395a corerun`ExecuteManagedAssembly(currentExeAbsolutePath=0x00007fff5fbffcb0, clrFilesAbsolutePath=0x0000000100104b30, managedAssemblyAbsolutePath=0x0000000100104b80, managedAssemblyArgc=0, managedAssemblyArgv=0x0000000000000000) + 3258 at corerun.cpp:308
    frame #13: 0x0000000100004340 corerun`main(argc=2, argv=0x00007fff5fbffbb8) + 1968 at corerun.cpp:406
    frame #14: 0x00007fff9512c5c9 libdyld.dylib`start + 1
    frame #15: 0x00007fff9512c5c9 libdyld.dylib`start + 1
(lldb) disas -c 5 -s $rip-6
   0x11b54a7af:  callq  0x10157a7d0               ; IL_Throw(Object*) at jithelpers.cpp:4993
   0x11b54a7b4:  int3   
-> 0x11b54a7b5:  nop    
   0x11b54a7b6:  leaq   -0x8(%rbp), %rsp
   0x11b54a7ba:  popq   %rbx
(lldb) 
```

The int3 outside lldb is causing us to fail with Trace/BPT trap: 5 since its not handled.

**Comment by sergiy-k** [comment]
@kangaroo I fixed RtlRestoreContext for OSX. With that change I can can complete the first test:
Testing A
Throwing an exception
A.finally()
Test A: success! Caught exception!
System.Exception: Did you catch it?
   at Program.A()
   at Program.Main()
Main.finally().
Testing B

But then it crashes in jit/flowgraph.cpp (L4939):
if ((opts.instrCount - ldStCount) < 4 || ((double)ldStCount/(double)opts.instrCount) > .90)

with an invalid instruction exception (0xc000001d, ExceptionAddress = 0x00000001017a5714). Here is disassembled code (please note this is NOT jitted code, it's C++ code compiled by llvm):
0x1017a570d:  89 c2                    movl   %eax, %edx
0x1017a570f:  f2 48 0f 2a d2           cvtsi2sdq %rdx, %xmm2
0x1017a5714:  f2 0f 5e ca              divsd  %xmm2, %xmm1
0x1017a5718:  66 0f 2e c8              ucomisd %xmm0, %xmm1

I need to look at Intel's developer manual because I don't remember requirements for these instructions.

**Comment by kangaroo** [comment]
@sergiy-k I was getting much further than that when I was testing, and wan't getting that crash.  Let me rebase your branch and see.

**Comment by kangaroo** [comment]
@sergiy-k I'm not getting the crash in flow graph, but after "Testing B" its busy looping.in the unwinder:

```
* thread #1: tid = 0x1470e0, 0x00000001019f495e libcoreclr.dylib`PAL_VirtualUnwind(context=0x00007fff5fbfd6d0, contextPointers=0x0000000000000000) + 46 at seh-unwind.cpp:128, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x00000001019f495e libcoreclr.dylib`PAL_VirtualUnwind(context=0x00007fff5fbfd6d0, contextPointers=0x0000000000000000) + 46 at seh-unwind.cpp:128
    frame #1: 0x000000010140721e libcoreclr.dylib`Thread::VirtualUnwindToFirstManagedCallFrame(pContext=0x00007fff5fbfd6d0) + 78 at stackwalk.cpp:782
    frame #2: 0x00000001015225e2 libcoreclr.dylib`DispatchManagedException(ex=0x000000011b810878) + 130 at exceptionhandling.cpp:4383
    frame #3: 0x00000001013ef1ff libcoreclr.dylib`PreStubWorker(pTransitionBlock=0x00007fff5fbfe2f8, pMD=0x00000001000e48a8) + 1567 at prestub.cpp:1003
    frame #4: 0x000000010167967b libcoreclr.dylib`JIT_WriteBarrier_Debug_End + 89 at jithelpers_slow.S:352
    frame #5: 0x000000011b547738
    frame #6: 0x0000000101679089 libcoreclr.dylib`CallDescrWorkerInternal + 125 at calldescrworkeramd64.S:902
    frame #7: 0x00000001014ad7e2 libcoreclr.dylib`CallDescrWorkerWithHandler(pCallDescrData=0x00007fff5fbfe570, fCriticalCall=0) + 482 at callhelpers.cpp:88
    frame #8: 0x00000001014aea7a libcoreclr.dylib`MethodDescCallSite::CallTargetWorker(this=0x00007fff5fbfe790, pArguments=0x00007fff5fbfe718) + 2426 at callhelpers.cpp:627
    frame #9: 0x000000010143da9d libcoreclr.dylib`MethodDescCallSite::Call(this=0x00007fff5fbfe790, pArguments=0x00007fff5fbfe718) + 29 at callhelpers.h:421
    frame #10: 0x0000000101112104 libcoreclr.dylib`RunMain(this=0x00007fff5fbfe8a8, pParam=0x00007fff5fbfe940)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const::'lambda'(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*)::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const + 580 at assembly.cpp:2540
    frame #11: 0x000000010110f45c libcoreclr.dylib`RunMain(this=0x00007fff5fbfe930, __EXparam=0x00007fff5fbfe940)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*)::Param*) const + 76 at assembly.cpp:2558
    frame #12: 0x000000010110f2f3 libcoreclr.dylib`RunMain(pFD=0x00000001000e4950, numSkipArgs=1, piRetVal=0x00007fff5fbfea58, stringArgs=0x00007fff5fbfedb0) + 323 at assembly.cpp:2558
    frame #13: 0x000000010110f628 libcoreclr.dylib`Assembly::ExecuteMainMethod(this=0x000000010021cc18, stringArgs=0x00007fff5fbfedb0) + 328 at assembly.cpp:2666
    frame #14: 0x00000001011923ff libcoreclr.dylib`CorHost2::ExecuteAssembly(this=0x000000010020f3c8, dwAppDomainId=1, pwzAssemblyPath=0x000000010021b0f8, argc=0, argv=0x0000000000000000, pReturnValue=0x00007fff5fbff2fc) + 1167 at corhost.cpp:1309
    frame #15: 0x0000000101004ce7 libcoreclr.dylib`ExecuteAssembly(exePath=0x00007fff5fbffcb0, coreClrPath=0x0000000100104c30, appDomainFriendlyName=0x0000000100007d80, propertyCount=4, propertyKeys=0x00007fff5fbff770, propertyValues=0x00007fff5fbff750, argc=0, argv=0x0000000000000000, managedAssemblyPath=0x0000000100104b80, exitCode=0x00007fff5fbff2fc) + 1655 at unixinterface.cpp:185
    frame #16: 0x000000010000395a corerun`ExecuteManagedAssembly(currentExeAbsolutePath=0x00007fff5fbffcb0, clrFilesAbsolutePath=0x0000000100104b30, managedAssemblyAbsolutePath=0x0000000100104b80, managedAssemblyArgc=0, managedAssemblyArgv=0x0000000000000000) + 3258 at corerun.cpp:308
    frame #17: 0x0000000100004340 corerun`main(argc=2, argv=0x00007fff5fbffbb8) + 1968 at corerun.cpp:406
    frame #18: 0x00007fff9512c5c9 libdyld.dylib`start + 1
```

**Comment by sergiy-k** [comment]
@kangaroo Yes, that's the crash I'm talking about. But the state you are seeing is just a consequence of the problem that I described. If you put a break point in SEHRaiseException, you'll see that during test B there is a hardware exception that happens in flow graph. Then PAL_DispatchException is executed and in its turn it calls SEHRaiseException which throws a PAL_SEHException. This is a scenario that we don't support yet.

So when PAL_SEHException is thrown in this case, it gets caught in PreStubWorker and then DispatchManagedException is called. You are right that the code got stuck in an infinite loop in VirtualUnwindToFirstManagedCallFrame because PAL_VirtualUnwind keeps returning the same address. Maybe there is a bug in our assembly routines that call PreStubWorker and they don't provide correct unwind info? I need to double check that and the Rip that is returned by PAL_VirtualUnwind.

**Comment by sergiy-k** [comment]
@kangaroo When you said "I was getting much further than that when I was testing, and wan't getting that crash." how much further did you get? Could you please clarify? Thank you.

**Comment by kangaroo** [comment]
@sergiy-k previously when I was using CONTEXT_SetThreadContextOnPort it was running until the int3 in the previous comment.

**Comment by kangaroo** [comment]
@sergiy-k We're getting an EXC_ARITHMETIC on the divsd instruction, because the contents are:

```
      xmm1 = {0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x40 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00}
      xmm2 = {0x00 0x00 0x00 0x00 0x00 0x00 0x2e 0x40 0x00 0x00 0x00 0xff 0x00 0xff 0x00 0xff}
```

**Comment by kangaroo** [comment]
Ok, so the issue is that we clobber mxcsr with a bad value when restoring the thread context.

```
#if 0
State.__fpu_mxcsr = lpContext->FltSave.MxCsr;
State.__fpu_mxcsrmask = lpContext->FltSave.MxCsr_Mask; // note: we don't save the mask for x86
#endif         
```

Commenting out like that allows the test to run.  I haven't dug into the capture side yet to see whats wrong.

**Comment by kangaroo** [comment]
That will get you running to the state I had earlier, where all tests run but we hit a int3 in JIT code.  Haven't figured out whats emitting it yet.

**Comment by kangaroo** [comment]
Test D is the only one failing, A, B, C and E are ok.  It seems to be unwinding back to the throw site, and an int3 is there (I'm guessing intentionally by the JIT) to stop continued control flow after a throw?

**Comment by kangaroo** [comment]
Ok, so for Test D, when we get into RtlRestoreContext, we switch to the correct $rip (site of IL_Throw + 2, one for int3 and one for nop).  The epilogue looks like:

```
(lldb) disas
-> 0x11bd3b7b4:  int3   
   0x11bd3b7b5:  nop    
   0x11bd3b7b6:  leaq   -0x8(%rbp), %rsp
   0x11bd3b7ba:  popq   %rbx
   0x11bd3b7bb:  popq   %rbp
   0x11bd3b7bc:  retq   
```

We set $rip to 0x11bd3b7b6, but the stack is incorrect.  So when we get to the retq on 0x11bd3b7bc the $ra points to 0x11bd3b7b4 instead of one managed frame up:

```
(lldb) bt
* thread #1: tid = 0x18dfdb, 0x000000011bd3b7b4, queue = 'com.apple.main-thread', stop reason = instruction step into
  * frame #0: 0x000000011bd3b7b4
    frame #1: 0x000000011bd3874f
    frame #2: 0x0000000101679089 libcoreclr.dylib`CallDescrWorkerInternal + 125 at calldescrworkeramd64.S:902
    frame #3: 0x00000001014ad7e2 libcoreclr.dylib`CallDescrWorkerWithHandler(pCallDescrData=0x00007fff5fbfe530, fCriticalCall=0) + 482 at callhelpers.cpp:88
    frame #4: 0x00000001014aea7a libcoreclr.dylib`MethodDescCallSite::CallTargetWorker(this=0x00007fff5fbfe750, pArguments=0x00007fff5fbfe6d8) + 2426 at callhelpers.cpp:627
    frame #5: 0x000000010143da9d libcoreclr.dylib`MethodDescCallSite::Call(this=0x00007fff5fbfe750, pArguments=0x00007fff5fbfe6d8) + 29 at callhelpers.h:421
    frame #6: 0x0000000101112104 libcoreclr.dylib`RunMain(this=0x00007fff5fbfe868, pParam=0x00007fff5fbfe900)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const::'lambda'(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*)::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const + 580 at assembly.cpp:2540
    frame #7: 0x000000010110f45c libcoreclr.dylib`RunMain(this=0x00007fff5fbfe8f0, __EXparam=0x00007fff5fbfe900)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*)::Param*) const + 76 at assembly.cpp:2558
    frame #8: 0x000000010110f2f3 libcoreclr.dylib`RunMain(pFD=0x00000001000e4950, numSkipArgs=1, piRetVal=0x00007fff5fbfea18, stringArgs=0x00007fff5fbfed70) + 323 at assembly.cpp:2558
    frame #9: 0x000000010110f628 libcoreclr.dylib`Assembly::ExecuteMainMethod(this=0x0000000100312788, stringArgs=0x00007fff5fbfed70) + 328 at assembly.cpp:2666
    frame #10: 0x00000001011923ff libcoreclr.dylib`CorHost2::ExecuteAssembly(this=0x0000000100109bd8, dwAppDomainId=1, pwzAssemblyPath=0x0000000100402958, argc=0, argv=0x0000000000000000, pReturnValue=0x00007fff5fbff2bc) + 1167 at corhost.cpp:1309
    frame #11: 0x0000000101004ce7 libcoreclr.dylib`ExecuteAssembly(exePath=0x00007fff5fbffcb0, coreClrPath=0x0000000100104c30, appDomainFriendlyName=0x0000000100007d80, propertyCount=4, propertyKeys=0x00007fff5fbff730, propertyValues=0x00007fff5fbff710, argc=0, argv=0x0000000000000000, managedAssemblyPath=0x0000000100104b80, exitCode=0x00007fff5fbff2bc) + 1655 at unixinterface.cpp:185
    frame #12: 0x000000010000395a corerun`ExecuteManagedAssembly(currentExeAbsolutePath=0x00007fff5fbffcb0, clrFilesAbsolutePath=0x0000000100104b30, managedAssemblyAbsolutePath=0x0000000100104b80, managedAssemblyArgc=0, managedAssemblyArgv=0x0000000000000000) + 3258 at corerun.cpp:308
    frame #13: 0x0000000100004340 corerun`main(argc=2, argv=0x00007fff5fbffb78) + 1968 at corerun.cpp:406
    frame #14: 0x00007fff9512c5c9 libdyld.dylib`start + 1
```

**Comment by sergiy-k** [comment]
@kangaroo Thanks a lot! Yes, I saw that too when I debugged it in the morning (I'm sorry for not posting my finding here - I just did not have time to do that because I had to run to do some errands). The problem is that when we call FixNonvolatileRegisters (vm\exceptionhandling.cpp) at the end of the second pass before resuming execution, it updates Rpb in the ContextRecord with an incorrect value (the value it sets is equal to the Rbp of IL_Throw but this may be just a coincidence). So we end up with an Rbp value that is less than Rsp which is clearly wrong. I still need to debug what happens in FixNonvolatileRegisters on OSX only (may a problem with a stack walker somewhere?). I have double checked that on Linux the ContextRecord is populated with the correct values.

**Comment by sergiy-k** [comment]
@kangaroo I’ve merged your PR for CONTEXT_FULL. Thank you! One of the reason it was not done from the beginning is that we currently don’t have this functionality implemented on Linux. Currently on Linux you can get only (CONTEXT_CONTROL | CONTEXT_INTEGER) and we cannot restore other registers. That’s why I opened https://github.com/dotnet/coreclr/issues/360 since it can be taken care by a separate PR.

**Comment by kangaroo** [comment]
@sergiy-k No need to apologize, it seems we did basically the same debugging before running out.  I just dumped my findings here in case someone else picked it up or that I had notes :)

Yeah, I followed the context issue on Linux, but I figured the complete capture is better on OS X, so we don't trash the fp control regs with random data :)

I havn't had a chance to look into the stack walker, and why we end up with the wrong $sp yet either.  I might have some time tonight, but I'm still getting up to speed with the windows-isms :+1:

**Comment by kangaroo** [comment]
@sergiy-k The rbp/rsp problem is probably the OSXTODO in seh-unwind.cpp.  We don't have support for libunwind getting spill slots on OSX.  I think @landonf has libtinyunwind which he offered under MIT/X11 which has support for this?

**Comment by kangaroo** [comment]
Yeah, thats the issue.  This patch:

https://gist.github.com/kangaroo/03fb9f5c167bca26a94a

Makes the test work, but obviously leaks memory.  If we don't want to bring in external code, I suppose we could refactor PAL_VirtualUnwind to allocate the context pointer store locations, and free after we de-reference, right before unwind?

That would be uglier code, but platform portable.

**Comment by sergiy-k** [comment]
@kangaroo Awesome! At least we now know where the problem is and that we need to fix GetContextPointer for OSX to make managed exceptions work. I took a look at how GetContextPointer and PAL_VirtualUnwind are used and I think it should not be too hard to make it work. But I do agree with you that it won't be super pretty. :) Once again thank you!
I also I agree that we should explore the option with libtinyunwind. @janvorli let's talk about it when you are back to work.

**Comment by kangaroo** [comment]
@sergiy-k :) I'm going to park this issue from my side until you and @janvorli talk given our discussions.  Let me know how you guys want to proceed and I'm happy to help integrate libtinyunwind, or refactor to support unw_step+free.

**Comment by janvorli** [comment]
@kangaroo @sergiy-k I'm back from my vacation, but I still need to go through a lot of emails, so just a quick note. Instead of allocating the storage for the register value pointed to by the context pointer, we can point it to the corresponding register member in the context. 
Anyways, this would still be just a workaround, since we will need the real context pointers for the GC. So the final solution should be either using the library from @landonf or doing what @jkotas has suggested - pinning the objects pointed to by registers which would remove the need for the context pointers. But maybe the issue with the rsp / rbp you were hitting would still need special handling. I didn't have a chance to look at the source code now.

**Comment by jkotas** [comment]
LGTM (as intermediate checkpoint)

**Comment by janvorli** [comment]
@kangaroo I can see that the `CONTEXT_GetThreadContextFromPort` uses incorrect testing of the `CONTEXT_xxx` flags. The problem is that they are not really flags, as you can see from their Amd64 definition below.
```c
#define CONTEXT_AMD64   0x100000

#define CONTEXT_CONTROL (CONTEXT_AMD64 | 0x1L)
#define CONTEXT_INTEGER (CONTEXT_AMD64 | 0x2L)
#define CONTEXT_SEGMENTS (CONTEXT_AMD64 | 0x4L)
#define CONTEXT_FLOATING_POINT  (CONTEXT_AMD64 | 0x8L)
#define CONTEXT_DEBUG_REGISTERS (CONTEXT_AMD64 | 0x10L)
```
So tests like `if (lpContext->ContextFlags & CONTEXT_FLOATING_POINT)` always pass no matter which `CONTEXT_xxx` flag is set, since the `CONTEXT_AMD64` is always set.

`CONTEXT_SetThreadContextOnPort` has the same issue and I guess it might exist at other OSX related places too, I wonder if that could  cause the issue you had with restoring FP state.

**Comment by janvorli** [comment]
@kangaroo I've forgotten to mention that after I have implemented full context restoring in the `RtlRestoreContext` for Linux, I started to get the same floating point issue. And the reason was a problem in the `CONTEXT_GetThreadContext` for Linux which was not getting the FP registers into the context even though the caller requested them. Thus the `lpContext->ContextFlags` didn't match the obtained context.

**Comment by kangaroo** [comment]
@janvorli good catch, that probably was the case.  I don't think it functionally matters now, since we always restore full fp and int states, but its probably a good item to split off into an issue for future clean up.

Speaking of which, how do we want to handle the context registers piece?  There is some useful discussion in this (now closed) PR.  Should we split a new issue for that and refer back?

**Comment by janvorli** [comment]
@kangaroo After looking at the usage of the context pointers in the FixNonvolatileRegisters, I believe the best solution is to get rid of the need to use context pointers on both Linux / OSX and pin the managed objects pointed to by registers instead. The FixNonvolatileRegisters was fixing register values in the context to ensure that if GC has moved the respective objects, the values in the context itself are up to date. So pinning the objects removes the need for this update as well.
The benefit is that we won't need any additional library on OSX and if we move to llvm's libunwind on Linux later (provided they make it a standalone library as expected), we will be fine too.
I've created issue #387 to represent that work.

**Comment by kangaroo** [comment]
@janvorli Sounds good to me, and I agree -- thats a much cleaner implementation.

**Pull request merged**